bitnami/redis/templates/configmap-scripts.yaml

```diff
@@ -294,28 +294,79 @@ data:
     #!/bin/bash
 
     . /opt/bitnami/scripts/libvalidations.sh
+    
+    run_sentinel_command() {
+      if is_boolean_yes "$REDIS_SENTINEL_TLS_ENABLED"; then
+        redis-cli -h ${REDIS_SERVICE} -p {{ .Values.sentinel.port }} --tls --cert ${REDIS_SENTINEL_TLS_CERT_FILE} --key ${REDIS_SENTINEL_TLS_KEY_FILE} --cacert ${REDIS_SENTINEL_TLS_CA_FILE} sentinel "$@"
+      else
+        redis-cli -h ${REDIS_SERVICE} -p {{ .Values.sentinel.port }} sentinel "$@"
+      fi
+    }
+    am_i_the_master() {
+      REDIS_SENTINEL_INFO=($(run_sentinel_command get-master-addr-by-name "{{ .Values.sentinel.masterSet }}"))
+      REDIS_MASTER_HOST="${REDIS_SENTINEL_INFO[0]}"
 
-    REDIS_SERVICE="{{ include "redis.fullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.clusterDomain }}"
+      [[ "$REDIS_MASTER_HOST" == "$(hostname -i)" ]]
+    }
 
+    REDIS_SERVICE="{{ include "redis.fullname" . }}.{{ .Release.Namespace }}.svc.{{ .Values.clusterDomain }}"
+    
+    if [[ -n "$REDIS_PASSWORD" ]]; then
+      export REDISCLI_AUTH="$REDIS_PASSWORD"
+    fi
     if [[ -n "$REDIS_PASSWORD_FILE" ]]; then
       password_aux=$(cat "$REDIS_PASSWORD_FILE")
-      export REDIS_PASSWORD="$password_aux"
+      export REDISCLI_AUTH="$password_aux"
     fi
 
-    if is_boolean_yes "$REDIS_SENTINEL_TLS_ENABLED"; then
-      sentinel_info_command="redis-cli {{- if .Values.usePassword }} -a ${REDIS_PASSWORD} {{- end }} -h ${REDIS_SERVICE} -p {{ .Values.sentinel.port }} --tls --cert ${REDIS_SENTINEL_TLS_CERT_FILE} --key ${REDIS_SENTINEL_TLS_KEY_FILE} --cacert ${REDIS_SENTINEL_TLS_CA_FILE} sentinel get-master-addr-by-name {{ .Values.sentinel.masterSet }}"
+    if am_i_the_master; then
+      run_sentinel_command failover "{{ .Values.sentinel.masterSet }}"
+      for i in $(seq 10); do
+        if ! am_i_the_master; then
+          exit 0
+        fi
+        echo "Waiting for failover to finish"
+        sleep 1
+      done
```

Comment:
When the loop finishes, the failover is not necessarily complete. Also, I think the second, if will never be executed since the for, will only be executed when you are the master right?
I would use something like:
``` suggestion
    run_failover () {
        run_sentinel_command failover "{{ .Values.sentinel.masterSet }}"
    }
    if ! retry_while run_failover; then
        error "Failed to run the failover"
        exit 1
    fi
```

Reply:
You are correct that the loop can end before we failover successfuly. But there is not much we can do about that. We cannot take longer than terminationGracePeriodSeconds (by default 30s) otherwise we get sigkilled. Also if the failover takes more than 10s you likely have some other problem elsewhere in the cluster that we cannot do anything about in the prestop hook.

The second if will be executed. You are correct that we enter this section of code only if we are master but once we execute the failover command on line 323 the process to move master to other node starts. It usually finishes within a second or two and after that the am_i_the_master function will return false and we can exit the wait loop early to make the pod shutdown as fast as possible.

Reply:
Your suggestion doesn't make much sense to me. It runs the failover command multiple times but that is really undesirable. Because each time you run it you swtich the master again and again. What we want to do is switch the master once (to get it away from the terminating pod) and then wait a bit until the process finishes.

Reply:
>It runs the failover command multiple times but that is really undesirable. 

I agree, I didn't notice that when writing the suggestion. Isn't there any way to check if the failover has finished? Maybe checking if the master has changed? If we find the way to do that check we can run the failover once, and use the `retry_while` with that new check command. I am not comfortable with the idea of leaving the sigkill in hands of the termination grace period since it is a configurable parameter and the `10` seconds will be a hardcoded number, IMHO it will become unconsistent.

Reply:
But that is precisely what I am doing on line 325 (`if ! am_i_the_master; then`). This checks if the current pod is still the master or not. If it is not the master anymore we terminate the loop and exit. I agree I could use the retry_while but I would need to know the definiton or some documentation for it so I know how exactly it behaves. Could you please point me to the function definition or documentation for the `retry_while` helper?

> I am not comfortable with the idea of leaving the sigkill in hands of the termination grace period since it is a configurable parameter and the 10 seconds will be a hardcoded number, IMHO it will become inconsistent.
This script works on the "best effort" basis. It tries to migrate the master away for 10s and then gives up and lets the pod terminate. I don't think anything more can be accomplished in the pre-stop hook. You are correct that the 10s is an arbitrary wait time but I am not sure what else to do. I definitely do not want to exceed the termination grace period and that is why I capped the wait to 10s. But if there is a way to get the current value of the termination grace period we could cap this proces to that minus 5s or something like that. That would make sure we exit quickly enough even when someone changes the grace period (but I do not think this chart exposes the grace period as a variable to be changed)

Reply:
Ah, just saw your other comment. I will check the docs and attempt to rework the loop with retry_while. Thanks!

Reply:
I believe this problem is resolved in the latest version of the PR